test/widgets/test_notify.py

```diff
@@ -169,3 +170,147 @@ def test_notifications(manager_nospawn, minimal_conf_noscreen, monkeypatch):
     # Show the display
     obj.display()
     assert obj.info()["text"] == MESSAGE_3.format(colour=LOW)
+
+
+def test_capabilities():
+    # Default capabilities are "body" and "actions"
+    widget_with_actions = notify.Notify()
+    assert widget_with_actions.capabilities == {"body", "actions"}
+
+    # If the user chooses not to have actions, the capabilities
+    # are adjusted accordingly
+    widget_no_actions = notify.Notify(action=False)
+    assert widget_no_actions.capabilities == {"body"}
+
+
+@pytest.mark.skipif(
+    shutil.which("notify-send") is None,
+    reason="notify-send not installed."
+    )
+@pytest.mark.usefixtures("dbus")
+def test_invoke_and_clear(manager_nospawn, minimal_conf_noscreen):
+
+    # We need to create an object to listen for signals from the qtile
+    # notification server. This needs to be created within the manager
+    # object so we rely on "eval" applying "exec".
+    handler = textwrap.dedent(
+        """
+        import asyncio
+
+        from libqtile.utils import add_signal_receiver
+
+        class SignalListener:
+            def __init__(sl):
+                sl.action_invoked = None
+                sl.notification_closed = None
+                global add_signal_receiver
+                global asyncio
+                asyncio.create_task(
+                    add_signal_receiver(
+                        sl.on_notification_closed,
+                        session_bus=True,
+                        signal_name="NotificationClosed"
+                    )
+                )
+
+                asyncio.create_task(
+                    add_signal_receiver(
+                        sl.on_action_invoked,
+                        session_bus=True,
+                        signal_name="ActionInvoked"
+                    )
+                )
+
+            def on_action_invoked(sl, msg):
+                sl.action_invoked = msg.body
+
+            def on_notification_closed(sl, msg):
+                sl.notification_closed = msg.body
+
+        self.signal_listener = SignalListener()
+        """
+    )
+
+    # Create and send a custom notification with a list of actions.
+    # `utils.send_notfication` is not an option here as it does not
+    # expose actions so we need a lower-level call
+    notification_with_actions = textwrap.dedent(
```

Comment:
Does this need to be run in the Qtile process or can it be run from the test process?

Reply:
That was my first idea. However, to listen for signals, I need the code to run in a loop. I didn't know how to run an asyncio loop in a test so I did it this way!

Reply:
If I run an event loop in my test, then I can't use `manager` objects as they can't do `asyncio.run` in an active loop. So, I think the option is either to use manager and my ugly hack, or run the test in a loop and try to execute the widget in that loop.

Reply:
Whichever is easiest :)

Reply:
Well, the easiest would be not to rewrite my code!

Reply:
Lol, then that! Overall LGTM